src/ds/buffer.h

```diff
@@ -8,6 +8,54 @@
 #include <utility>
 #include <vector>
 
+// The OArray (Owning Array) owns a buffer and provides a projection onto said
+// buffer via a pointer and a length.
+class OArray
+{
+public:
+  OArray(std::vector<uint8_t> d_) :
+    d(std::move(d_)),
+    data_(d_.data()),
+    size_(d_.size())
+  {}
+
+  OArray(const OArray& other) = delete;
+  OArray(OArray& other) = delete;
+  OArray& operator=(const OArray& rhs) = delete;
+  OArray& operator=(OArray& rhs) = delete;
+
+  OArray(OArray&& other)
+  {
+    data_ = other.data_;
+    size_ = other.size_;
+    d = std::move(other.d);
+  }
+  OArray& operator=(OArray&& other)
+  {
+    data_ = other.data_;
+    size_ = other.size_;
+    d = std::move(other.d);
+
+    return *this;
+  }
```

Comment:
I suggest we assert that data + size <= d.data() + d.size(), here and in the ctor. It'll macro away in release, and it might catch some stuff in debug.

Reply:
Also data >= d.data()